File: scripts/create_model.py
import torch
import torch.nn as nn
import os

def create_model():
    # Model dimensions
    embed_dim = 2048
    num_layers = 8
    num_heads = 8
    num_kv_heads = 2
    vocab_size = 128_256
    
    # Create state dict
    state_dict = {}
    
    # Token embedding
    state_dict['token_embedding.weight'] = torch.randn(vocab_size, embed_dim)
    
    # Transformer layers
    for i in range(num_layers):
        layer_prefix = f'layer_{i}'
        
        # Self attention
        state_dict[f'{layer_prefix}.self_attn.q_proj.weight'] = torch.randn(embed_dim, embed_dim)
        state_dict[f'{layer_prefix}.self_attn.q_proj.bias'] = torch.randn(embed_dim)
        
        state_dict[f'{layer_prefix}.self_attn.k_proj.weight'] = torch.randn(embed_dim, embed_dim)
        state_dict[f'{layer_prefix}.self_attn.k_proj.bias'] = torch.randn(embed_dim)
        
        state_dict[f'{layer_prefix}.self_attn.v_proj.weight'] = torch.randn(embed_dim, embed_dim)
        state_dict[f'{layer_prefix}.self_attn.v_proj.bias'] = torch.randn(embed_dim)
        
        state_dict[f'{layer_prefix}.self_attn.o_proj.weight'] = torch.randn(embed_dim, embed_dim)
        state_dict[f'{layer_prefix}.self_attn.o_proj.bias'] = torch.randn(embed_dim)
        
        # Layer norms
        state_dict[f'{layer_prefix}.norm1.weight'] = torch.ones(embed_dim)
        state_dict[f'{layer_prefix}.norm1.bias'] = torch.zeros(embed_dim)
        
        state_dict[f'{layer_prefix}.norm2.weight'] = torch.ones(embed_dim)
        state_dict[f'{layer_prefix}.norm2.bias'] = torch.zeros(embed_dim)
        
        # MLP
        intermediate_dim = embed_dim * 4
        state_dict[f'{layer_prefix}.mlp.w1.weight'] = torch.randn(intermediate_dim, embed_dim)
        state_dict[f'{layer_prefix}.mlp.w1.bias'] = torch.randn(intermediate_dim)
        
        state_dict[f'{layer_prefix}.mlp.w2.weight'] = torch.randn(embed_dim, intermediate_dim)
        state_dict[f'{layer_prefix}.mlp.w2.bias'] = torch.randn(embed_dim)
    
    # Final layer norm
    state_dict['norm.weight'] = torch.ones(embed_dim)
    state_dict['norm.bias'] = torch.zeros(embed_dim)
    
    # Output projection
    state_dict['output.weight'] = torch.randn(vocab_size, embed_dim)
    state_dict['output.bias'] = torch.randn(vocab_size)
    
    # Save weights
    os.makedirs('models', exist_ok=True)
    torch.save(state_dict, 'models/llama-1B.pth')
    print("Model weights saved to models/llama-1B.pth")

if __name__ == '__main__':
    create_model() 

File: src/bin/create_model.rs
use anyhow::Result;
use tch::{nn, Device, Kind, Tensor};
use csm::models::{Model, ModelConfig};

fn main() -> Result<()> {
    let config = ModelConfig {
        backbone_flavor: "llama-1B".to_string(),
        decoder_flavor: "default".to_string(),
        text_vocab_size: 128_256,
        audio_vocab_size: 1024,
        audio_num_codebooks: 8,
        max_seq_len: 2048,
        embed_dim: 2048,
    };

    let device = Device::Cpu;
    let mut vs = nn::VarStore::new(device);
    let root = vs.root();

    // Initialize token embeddings
    root.var_copy(
        "token_embedding_weight",
        &(&Tensor::randn(&[config.text_vocab_size as i64, config.embed_dim as i64], (Kind::Float, device)) * 0.02)
    );

    // Initialize transformer layers
    for i in 0..12 { // Fixed number of layers for llama-1B
        let layer_root = root.sub(&format!("layer_{}", i));
        
        // Layer norms
        layer_root.var_copy(
            "norm1_weight",
            &Tensor::ones(&[config.embed_dim as i64], (Kind::Float, device))
        );
        layer_root.var_copy(
            "norm1_bias",
            &Tensor::zeros(&[config.embed_dim as i64], (Kind::Float, device))
        );
        layer_root.var_copy(
            "norm2_weight",
            &Tensor::ones(&[config.embed_dim as i64], (Kind::Float, device))
        );
        layer_root.var_copy(
            "norm2_bias",
            &Tensor::zeros(&[config.embed_dim as i64], (Kind::Float, device))
        );

        // Self attention
        layer_root.var_copy(
            "self_attn_q_proj_weight",
            &(&Tensor::randn(&[config.embed_dim as i64, config.embed_dim as i64], (Kind::Float, device)) * 0.02)
        );
        layer_root.var_copy(
            "self_attn_k_proj_weight",
            &(&Tensor::randn(&[config.embed_dim as i64, config.embed_dim as i64], (Kind::Float, device)) * 0.02)
        );
        layer_root.var_copy(
            "self_attn_v_proj_weight",
            &(&Tensor::randn(&[config.embed_dim as i64, config.embed_dim as i64], (Kind::Float, device)) * 0.02)
        );
        layer_root.var_copy(
            "self_attn_o_proj_weight",
            &(&Tensor::randn(&[config.embed_dim as i64, config.embed_dim as i64], (Kind::Float, device)) * 0.02)
        );

        // MLP
        layer_root.var_copy(
            "mlp_w1_weight",
            &(&Tensor::randn(&[config.embed_dim as i64, 4096], (Kind::Float, device)) * 0.02)
        );
        layer_root.var_copy(
            "mlp_w2_weight",
            &(&Tensor::randn(&[4096, config.embed_dim as i64], (Kind::Float, device)) * 0.02)
        );
    }

    // Final layer norm
    root.var_copy(
        "norm_weight",
        &Tensor::ones(&[config.embed_dim as i64], (Kind::Float, device))
    );
    root.var_copy(
        "norm_bias",
        &Tensor::zeros(&[config.embed_dim as i64], (Kind::Float, device))
    );

    // Output projection
    root.var_copy(
        "output_weight",
        &(&Tensor::randn(&[config.audio_vocab_size as i64, config.embed_dim as i64], (Kind::Float, device)) * 0.02)
    );
    root.var_copy(
        "output_bias",
        &Tensor::zeros(&[config.audio_vocab_size as i64], (Kind::Float, device))
    );

    // Save the model
    vs.save("models/llama-1B.pth")?;
    println!("Model weights saved to models/llama-1B.pth");
    Ok(())
} 